### PR TITLE
[CARBONDATA-2482] Pass uuid while writing segment file if possible

### PR DESCRIPTION
**Pull request by dhatchayani**
Pass the uuid (segmentFileName) to the writeSegmentFile method file if possible.

**Problem:**
When the supporting tables depends on the segmentFileName of the main table, query is failing, as it is expected to be the same name as that of the main table. Mostly in case of merge index, segmentFile will be rewritten for that segment.

**Solution:**
Whenever the supporting table segmentFileName should be same as that main table, in that case we should pass the name as the UUID in the merge index flow instead of taking a new timestamp.

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [x] Testing done
        UTs are already there
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5884/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4730/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4920/

**Comment by dhatchayani** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5890/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4736/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4757/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/5912/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/4945/

**Comment by manishgupta88** [comment]
@dhatchayani ...I have few doubts
1. Why UUID is optional here? I can see from some methods it is passed as null and then System.currentTime is used as UUID
2. Please update the description and give detailed explanation  which clearly defines the purpose of this PR

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 1]
```diff
@@ -1085,11 +1085,11 @@ public static Long addDataIndexSizeIntoMetaEntry(LoadMetadataDetails loadMetadat
    * @return
    * @throws IOException
    */
-  public static String mergeIndexFilesinPartitionedSegment(String segmentId, CarbonTable table)
-      throws IOException {
+  public static String mergeIndexFilesinPartitionedSegment(String segmentId, String uuid,
+      CarbonTable table) throws IOException {
```
Comment (manishgupta88) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: make this as the first argument

**Comment by manishgupta88** [comment]
LGTM...can be merged once build passes

File: core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java [hunk 2]
```diff
@@ -145,9 +145,10 @@ private String writeMergeIndexFileBasedOnSegmentFile(
       }
     }
 
-    String uniqueId = String.valueOf(System.currentTimeMillis());
-    String newSegmentFileName =
-        SegmentFileStore.genSegmentFileName(segmentId, String.valueOf(uniqueId))
+    String uniqueId = (uuid.endsWith(CarbonTablePath.SEGMENT_EXT)) ?
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/writer/CarbonIndexFileMergeWriter.java]: Don't pass segment file name, just pass only uuid .

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5820/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7113/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5831/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5889/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/5843/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7182/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5958/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7200/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/5976/

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**